esp-hal-common/src/uart.rs

```diff
@@ -1586,41 +1636,26 @@ mod asynch {
         ///   To avoid this error, call this function more often.
         ///
         /// # Ok
-        /// When succesfull, returns the number of bytes written to buf.
+        /// When successful, returns the number of bytes written to buf.
         /// This method will never return Ok(0), unless buf.len() == 0.
         async fn read_async(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
             if buf.len() == 0 {
                 return Ok(0);
             }
 
             loop {
-                let mut futures: Vec<_, 4> = Vec::new();
-                futures
-                    .push(UartFuture::<T>::new(Event::RxFifoFull))
-                    .map_err(|_| ())
-                    .unwrap();
-                futures
-                    .push(UartFuture::<T>::new(Event::RxFifoOvf))
-                    .map_err(|_| ())
-                    .unwrap();
+                let mut events: Vec<RxEvent, 4> =
+                    Vec::from_slice(&[RxEvent::RxFifoFull, RxEvent::RxFifoOvf]).unwrap();
 
                 if self.at_cmd_config.is_some() {
-                    futures
-                        .push(UartFuture::<T>::new(Event::RxCmdCharDetected))
-                        .map_err(|_| ())
-                        .unwrap();
+                    events.push(RxEvent::RxCmdCharDetected).unwrap();
                 }
 
                 if self.rx_timeout_config.is_some() {
-                    futures
-                        .push(UartFuture::<T>::new(Event::RxFifoTout))
-                        .map_err(|_| ())
-                        .unwrap();
-                }
-
-                if select_slice(&mut futures).await.0 == Event::RxFifoOvf {
-                    return Err(Error::RxFifoOvf);
+                    events.push(RxEvent::RxFifoTout).unwrap();
                 }
+                let hej = UartRxFuture::<T>::new(&events);
+                hej.await;
```

Comment:
Before we checked whether an RxFifoOvf interrupt occurred. It seems you've removed this. Is there a reason for that?
UartRxFuture should probably only return the events that actually occurred, not all that were polled.

Reply:
It was broken as it was. It never emptied the buffer in case of overflow. And it can miss overflow as well. There is still work to be done to signal overflow from interrupt_handler to future. But I think we can manage that in a separate pull request? I will push an update and remove the "hej" ;) and remove the return type from the futures